libs/base/Control/Relation.idr

```diff
@@ -0,0 +1,12 @@
+||| Properties of binary relations
+module Control.Relation
+
+%default total
+
+public export
+interface Reflexive (ty : Type) (rel : ty -> ty -> Type) where
```

Comment:
What you're describing is possible in Idris 1, but it is not currently possible in Idris 2. As inference improves, I would be happy to get rid of redundancies like this. That said, it isn't the worst thing in the world to have a little extra explicitness. `Reflexive Nat LTE` can be read as: "`LTE` is `Reflexive` over `Nat`".

Reply:
Maybe you're right. But if it's impossible, it's all academic anyway.

Reply:
I think it is indeed possible, see the test case in #305

Reply:
Do you have any concrete suggestions for how to do that? I think what we would like to see is:

```idris
interface Reflexive (rel : ty -> ty -> Type) where
  reflexive : {x : ty} -> rel x x

Reflexive LTE where
  reflexive {x = Z} = LTEZero
  reflexive {x = S k} = LTESucc $ reflexive {x = k}
```

But that turns up this error:

```
Control/Relation.idr:12:31--12:33:While processing type of reflexive at Data/Nat.idr:77:1--81:1:
When unifying ?ty [no locals in scope] and Nat
Pattern variable ty unifies with:
	Nat
```

Reply:
That looks like a bug.

Reply:
I don't think so. I thought about putting `ty : Nat` in `{}` braces, but that is indeed a syntax error. However, omitting the argument altogether renders you unable to instantiate the interface with a concrete `ty` type, because Idris then demands it to be abstract (which seems reasonable). It's also consistent with behaviour for regular functions, where omitting a type argument means you can't refer to it in the definition.  I don't know how this was resolved in Idris1, but here I would like to be able to put `ty : Type` in `{}`, which again would be consistent with behaviour for functions. But that is probably a big change…